diffpy/snmf/subroutines.py

```diff
@@ -49,3 +50,35 @@ def objective_function(residual_matrix, stretching_factor_matrix, smoothness, sm
     component_matrix = np.asarray(component_matrix)
     return .5 * np.linalg.norm(residual_matrix, 'fro') ** 2 + .5 * smoothness * np.linalg.norm(
         smoothness_term @ stretching_factor_matrix.T, 'fro') ** 2 + sparsity * np.sum(np.sqrt(component_matrix))
+
+
+def get_stretched_component(stretching_factor, component, signal_length):
+    """Applies a stretching factor to a component signal.
+
+    Computes a stretched signal and reinterpolates it onto the original grid of points. Uses a normalized grid of evenly
+    spaced integers counting from 0 to signal_length (exclusive) to approximate values in between grid nodes. Once this
+    grid is stretched, values at grid nodes past the unstretched signal's domain are set to zero. Returns the
+    approximate values of x(r/a) from x(r) where x is a component signal.
+
+    Parameters
+    ----------
+    stretching_factor: float
+      The stretching factor of a component signal at a particular moment.
+    component: 1d array like
+      The calculated component signal without stretching or weighting. Has length N, the length of the signal.
+    signal_length: int
+      The length of the component signal.
+
+    Returns
+    -------
+    1d array of floats
+      The calculated component signal with stretching factors applied. Has length N, the length of the unstretched
+      component signal.
+
+    """
+    component = np.asarray(component)
+    normalized_grid = np.arange(signal_length)
+    spline = scipy.interpolate.InterpolatedUnivariateSpline(normalized_grid, component, k=1, ext=1)
```

Comment:
I was expecting something a bit more like this:
`ynew = np.interp(xnew, x, y)`

so we have an x-grid and the intensities on the x-grid, then we stretch the x-grid, then we just need to interpolate the stretched function onto the original grid, so I think it goes something like:
```
x_stretched = [x*factor for x in original_grid]
ynew = np.interp(original_grid, x_stretched, y)
```
which will put the new stretched function back onto the unstretched grid.

Does that work?  The code is much cleaner.

Reply:
This works. I will implement it. In my email to Ran, he said we are able to use a normalized grid which is just the a vector from 0 to the length of the signal.